src/workerd/api/node/zlib-util.c++

```diff
@@ -1,11 +1,373 @@
+  switch (err) {
+  case Z_OK:
+  case Z_BUF_ERROR:
+    if (stream.avail_out != 0 && flush == Z_FINISH) {
+      return constructError("unexpected end of file"_kj);
+    }
+  case Z_STREAM_END:
+    // normal statuses, not fatal
+    break;
+  case Z_NEED_DICT:
+    if (dictionary.empty())
+      return constructError("Missing dictionary"_kj);
+    else
+      return constructError("Bad dictionary"_kj);
+  default:
+    // something else.
+    return constructError("Zlib error");
+  }
+
+  return {};
+}
+
+kj::Maybe<CompressionError> ZlibContext::setDictionary() {
+  if (dictionary.empty()) {
+    return kj::none;
+  }
+
+  switch (mode) {
+  case ZlibMode::DEFLATE:
+  case ZlibMode::DEFLATERAW:
+    err = deflateSetDictionary(&stream, dictionary.begin(), dictionary.size());;
+    break;
+  case ZlibMode::INFLATERAW:
+    err = inflateSetDictionary(&stream, dictionary.begin(), dictionary.size());
+    break;
+  default:
+    break;
+  }
+
+  if (err != Z_OK) {
+    return constructError("Failed to set dictionary"_kj);
+  }
+
+  return kj::none;
+}
+
+bool ZlibContext::initializeZlib() {
+  if (initialized.lockExclusive()) {
+    return false;
+  }
+  switch (mode) {
+  case ZlibMode::DEFLATE:
+  case ZlibMode::GZIP:
+  case ZlibMode::DEFLATERAW:
+    err = deflateInit2(&stream, level, Z_DEFLATED, windowBits, memLevel, strategy);
+    break;
+  case ZlibMode::INFLATE:
+  case ZlibMode::GUNZIP:
+  case ZlibMode::INFLATERAW:
+  case ZlibMode::UNZIP:
+    err = inflateInit2(&stream, windowBits);
+    break;
+  default:
+    KJ_UNREACHABLE;
+  }
+
+  if (err != Z_OK) {
+    dictionary.clear();
+    mode = ZlibMode::NONE;
+    return true;
+  }
+
+  setDictionary();
+  *initialized.lockExclusive() = true;
+  return true;
+}
+
+kj::Maybe<CompressionError> ZlibContext::resetStream() {
+  bool initialized_now = initializeZlib();
+  if (initialized_now && err != Z_OK) {
+    return constructError("Failed to init stream before reset");
+  }
+  switch (mode) {
+  case ZlibMode::DEFLATE:
+  case ZlibMode::DEFLATERAW:
+  case ZlibMode::GZIP:
+    err = deflateReset(&stream);
+    break;
+  case ZlibMode::INFLATE:
+  case ZlibMode::INFLATERAW:
+  case ZlibMode::GUNZIP:
+    err = inflateReset(&stream);
+    break;
+  default:
+    break;
+  }
+
+  if (err != Z_OK) {
+    return constructError("Failed to reset stream"_kj);
+  }
+
+  return setDictionary();
+}
+
+void ZlibContext::work() {
+  bool initialized_now = initializeZlib();
+  if (initialized_now && err != Z_OK) {
+    return;
+  }
+  const Bytef* next_expected_header_byte = nullptr;
+
+  err = Z_OK;
+
+  // If the avail_out is left at 0, then it means that it ran out
+  // of room.  If there was avail_out left over, then it means
+  // that all the input was consumed.
+  switch (mode) {
+    case ZlibMode::DEFLATE:
+    case ZlibMode::GZIP:
+    case ZlibMode::DEFLATERAW:
+      err = deflate(&stream, flush);
+      break;
+    case ZlibMode::UNZIP:
+      if (stream.avail_in > 0) {
```

Comment:
This case is very complex. I would consider to extract it into its own function and invoke it from here.